presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregationIntoTableScan.java

```diff
@@ -35,54 +35,48 @@
 import io.prestosql.sql.planner.Symbol;
 import io.prestosql.sql.planner.iterative.Rule;
 import io.prestosql.sql.planner.plan.AggregationNode;
-import io.prestosql.sql.planner.plan.AggregationNode.GroupingSetDescriptor;
 import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.GroupIdNode;
 import io.prestosql.sql.planner.plan.ProjectNode;
 import io.prestosql.sql.planner.plan.TableScanNode;
 import io.prestosql.sql.tree.Expression;
 import io.prestosql.sql.tree.SymbolReference;
 
+import java.util.Collection;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 import static com.google.common.base.Verify.verify;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
 import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.matching.Pattern.typeOf;
 import static io.prestosql.sql.planner.plan.Patterns.aggregation;
 import static io.prestosql.sql.planner.plan.Patterns.source;
 import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toSet;
 
-public class PushAggregationIntoTableScan
+public abstract class PushAggregationIntoTableScan
```

Comment:
Using inheritance for code reuse is unnecessary (and undesirable). This could be restructured as two independent rules that delegate to some piece of common code for the core of the logic:

```
class PushAggregationIntoTableScanWithoutGroupingSets
        implements Rule<AggregationNode>
{
    private static final Pattern<AggregationNode> PATTERN = ...

    public Result apply(AggregationNode node, Captures captures, Context context) 
    {
        TableScanNode tableScan = captures.get(TABLE_SCAN);
        // extract group by symbols and mappings
        ...
       return pushAggregationIntoTableScan(...);
}
```

The variant for grouping sets would differ in the pattern and how it extracts the group by symbols and mappings

Reply:
Changed to delegation.